gensim/corpora/dictionary.py

```diff
@@ -148,9 +148,9 @@ def doc2bow(self, document, allow_update=False, return_missing=False):
 
         token2id = self.token2id
         if allow_update or return_missing:
-            missing = {w: freq for w, freq in iteritems(counter) if w not in token2id}
+            missing = sorted((x for x in iteritems(counter) if x[0] not in token2id), key=lambda x: (x[1], x[0]))
```

Comment:
Sorting by (freq, w), are you sure that this is correct?

Reply:
@menshikh-iv, that's a good point. Either way works and ensures reproducibility. It can be assumed that sorting integers is faster than strings, but then the branching for words with the same frequency seems to introduce overhead (see the figure, where `freq` indicates frequency-word sorting and `string` indicates word-frequency sorting, `i`, `f`, `c`, and `d` methods are as in the previous example).
![string-integer](https://user-images.githubusercontent.com/29557375/33101444-685df42a-cf6c-11e7-8869-718ab9f198b1.png)

Since the words in `token2id` guaranteed to be unique but the frequencies are not (and will likely overlap given patterns within natural language), sorting by token then frequency is probably the better choice but will vary on a dataset-by-dataset basis. What's you're preference?

Reply:
IMO `(w, freq)` better than `(freq, w)` (just because of the uniqueness of the tokens). For bigger corpus, the situation will be worse because frequencies obey [zipf's law](https://simple.wikipedia.org/wiki/Zipf%27s_law) -> more and more duplicates by freq in "tail" of distribution.

Reply:
Made the changes.